### PR TITLE
fix: file upload without buffering

### LINKED ISSUE #1529: Improve file add so it can handle gigs of data
This is related to https://github.com/ipfs/js-ipfs/issues/3029, however changes on the webui end would also be required. So this is the placeholder issue for that piece of work.

### PR DESCRIPTION
**Pull request by Gozala**
# Version 2

This pull request takes advantage of following changes

- [x] [Use native form data in ipfs-http-client](https://github.com/ipfs/js-ipfs/pull/3184)
- [x] [Upload / Download progress events](https://github.com/ipfs/js-ipfs-utils/pull/54)

Please note that https://github.com/ipfs/js-ipfs/pull/3184 is coming in `ipfs-http-client@45.0.1` and to observe the fix you'd need to modify dependency in `package.json` as follows. I do not think we need to block this on release since minor version bump will be picked up automatically.

```diff
diff --git a/package.json b/package.json
index d87f56c..e129beb 100644
--- a/package.json
+++ b/package.json
@@ -47,7 +47,7 @@
     "ip": "^1.1.5",
     "ipfs-css": "^1.2.0",
     "ipfs-geoip": "^4.1.0",
-    "ipfs-http-client": "^45.0.0",
+    "ipfs-http-client": "next",
     "ipfs-provider": "^1.0.0",
     "ipld-explorer-components": "1.6.0",
     "is-binary": "^0.1.0",

```

Also note that while pull request introducing upload / download progress events is still pending, we don't need to block this because:

1. Prior versions of ipfs-http-client just ignores those handlers.
2. Progress handling is not in place anyway.

Therefor we can land this and once ipfs-http-client starts emitting upload / download progress events they will just start emitting `FILES_WRITE_UPDATED` actions.

_Everything below is no longer reflects current state of the pull request, but keeping it around for history._

# Version 1 (Retaining for historical)

This pull request explores a ways to upload files without unnecessary buffering. High level overview of changes below:

- `ipfs.add` has a complex API that can take `AsyncIterable` of files with `AsyncIterable` contents. Optimize such API has several issues:
   1. Optimizing such API is very complex, as all the current input normalization would have to be thrown out of the window and each input type would require separate specialized code path. Simply put it would increase implementation complexity quite a bit.
   2. Even if optimizations are made (despite introduced complexity) it would be really easy to fall of happy path by mixing inputs that can't be optimized.

  For above reasons this draft pull in experimental `ipfs-lite-http-client` providing much more simplified `ipfs.add` functionality that only accepts inputs optimal for web. It is not here to stay, but here to make a case that `ipfs.add` API is inadequate for use cases like this. 

2. `filesToStreams` function used to turn DOM [`File`](https://developer.mozilla.org/en-US/docs/Web/API/File) into structure with `content`s of it represented as a stream. Which in turn than had to be buffered by an `ipfs-http-client`. This function is gone now and `File` instances are used instead. **However I'm not confident there is no code paths that assume old file like structures instead of `File` instances.

3. I can successfully upload files with multiple gigs in size, it does take quite a bit time and I see no progress reports. `ipfs-lite-http-client` does report progress updates and I can see those being called however I can't seem to figure out where dispatch of `FILES_WRITE_UPDATED` leads to. I suspect handler of that action doesn't get what it expects and that is why progress doesn't get reported till the very end. Help here would be welcome.

might fix #1529

### CODE REVIEW

**Comment by Gozala** [comment]
> I can successfully upload files with multiple gigs in size, it does take quite a bit time and I see no progress reports. ipfs-lite-http-client does report progress updates and I can see those being called however I can't seem to figure out where dispatch of FILES_WRITE_UPDATED leads to. I suspect handler of that action doesn't get what it expects and that is why progress doesn't get reported till the very end. Help here would be welcome.

per @olizilla import / add button on the files view listened for `FILES_WRITE_UPDATED` events to show a progress bar-ish animation, but that no longer seems to be the case.

@rafaelramalho19 any idea what  am I doing wrong in my pull request, that causes the import view not to show until upload is fully complete ?

**Comment by Gozala** [comment]
Here is the timing I observe for 2gig file add

image

Which is ~45 seconds, that feels like a lot (especially when there is no visible UI feedback through all this time).

**Comment by Gozala** [comment]
Contrast with adding via CLI is massive extra 43 seconds

```
time ipfs add 2g.bloat --quiet
QmTbKT35oQzKgR34LUHaPdtrFrphjxiQJLrz6QMyU4kQNC
ipfs add 2g.bloat --quiet  0.71s user 1.49s system 13% cpu 16.138 total
```

I do not exactly know what is the problem here, but clearly this needs investigation.

**Comment by Gozala** [comment]
Checked also with curl which takes around ~3secs, which is kind of throughput I would like to see with webUI.

```
time curl -X POST -F file=@2g.bloat http://127.0.0.1:5001/api/v0/add\?stream-channels\=true\&pin\=false\&progress\=true\&wrap-with-directory\=false
{"Name":"2g.bloat","Bytes":2147483648}
{"Name":"2g.bloat","Hash":"QmTbKT35oQzKgR34LUHaPdtrFrphjxiQJLrz6QMyU4kQNC","Size":"2147994442"}
curl -X POST -F file=@2g.bloat   0.60s user 2.92s system 12% cpu 28.167 total
```

**Comment by Gozala** [comment]
Wrote a simple node server that just calculates number of bytes received and writes back chunks back
`{ Bytes: bytes, Name: "2g.bloat" }`

Which leads to upload times between 10s to 12s

image

With that 19s to do actual IPFS add doesn't seem all that bad. I also see from server output that browser is streaming data in chunks, that being said it is still unclear why it is taking this long.

**Comment by Gozala** [comment]
- [ ] It would be nice to add [server timing API](https://web.dev/custom-metrics/?utm_source=devtools#server-timing-api) into IPFS HTTP API.

**Comment by Gozala** [comment]
I did bit more investigation with my node server and it appears that

Chrome uploads 2gigs file in 108237 chunks (~19kb per chunk)
curl uploads same file in 32945 chunks (65kb per chunk)

This leads me believe that difference in chunking could lead to observed throughput difference. I am unaware if there is any way to control chunking of XHR upload, so I'm out of ideas.

**Comment by Gozala** [comment]
> @rafaelramalho19 any idea what am I doing wrong in my pull request, that causes the import view not to show until upload is fully complete ?

Relevant pointer to https://github.com/ipfs-shipyard/ipfs-webui/pull/1495 from @rafaelramalho19

**Comment by Gozala** [comment]
Blocked on https://github.com/ipfs/js-ipfs-utils/pull/54

**Comment by Gozala** [comment]
Something is not working, dropping a file seems to start spinning CPU and tab is unresponsive. I think there's non terminating recursion somewhere. 😭

**Comment by rafaelramalho19** [comment]
Looks good! I can finally upload 10gb files on both MacOS and Windows 🙏 

After this gets unblocked feel free to merge 😄

File: src/bundles/files/actions.js [hunk 1]
```diff
@@ -283,7 +283,11 @@ export default () => ({
      */
     async (ipfs, files, root, id, { dispatch }) => {
       files = files.filter(f => !IGNORED_FILES.includes(basename(f.path)))
-      const totalSize = files.reduce((prev, { size }) => prev + size, 0)
+      const uploadSize = files.reduce((prev, { size }) => prev + size, 0)
+      // Just estimate download size to be around 10% of upload size.
+      const downloadSize = uploadSize * 10 / 100
```
[the remark this answers is not in the document]
Comment (Gozala) [reply src/bundles/files/actions.js:288]: It refers to the response body estimate. Which would translate to progressbar speeding up towards the end.

**Comment by Gozala** [comment]
> After this gets unblocked feel free to merge 😄

Can you please elaborate what you mean ? We can land this now and once new ipfs-http-client is published things would actually stop buffering. Alternatively we could wait for ipfs-http-client to be published first and then bump dependency version. However since it's going to minor version update I don't think it's worth waiting.

**Comment by lidel** [comment]
@Gozala was new ipfs-http-client published with the changes we were waitig for? 
If so, can you switch  this PR to it?

**Pull request merged**